### PR TITLE
DATAUP-747: XSV form widgetification!

### PR DESCRIPTION
**Pull request by ialarmedalien**
# Description of PR purpose/changes

- Widgetised XSV form
- Fixed bug involving display values not being populated in the CSV/TSV/Excel data submission

# Jira Ticket / Issue #

Related Jira ticket: https://kbase-jira.atlassian.net/browse/DATAUP-747
- [x] Added the Jira Ticket to the title of the PR (e.g. `DATAUP-69 Adds a PR template`)

# Testing Instructions
* Details for how to test the PR:
- [x] Tests pass locally and in GitHub Actions
- [x] Changes available by spinning up a local narrative, cracking open a bulk import cell, and then clicking on that 'generate template' button.

# Dev Checklist:

- [x] My code follows the guidelines at https://sites.google.com/lbl.gov/trussresources/home?authuser=0
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] (JavaScript) I have run Prettier and ESLint on changed code manually or with a git precommit hook

### CODE REVIEW

File: nbextensions/bulkImportCell/tabs/xsvGenerator.js [hunk 1]
```diff
@@ -1,112 +1,99 @@
...
-            'output_directory', // file path
-        ],
-        specs: {
-            types: {
-                original: {
-                    advanced: 0,
-                    allow_multiple: 1,
-                    default_values: [
-                        '', // current type
+    const xsvAppSpec = {
```
Comment (ialarmedalien) [inline nbextensions/bulkImportCell/tabs/xsvGenerator.js:24]: convert this structure into a genuine app spec

File: nbextensions/bulkImportCell/tabs/xsvGenerator.js [hunk 2]
```diff
@@ -1,112 +1,99 @@
...
+                allow_multiple: 0,
+                default_values: [
+                    '', // current type
+                ],
+                description: 'File types to generate templates for',
+                disabled: 0,
+                dropdown_options: {
+                    multiselection: 1,
+                    options: [
+                        // map file type to display name
```
Comment (ialarmedalien) [inline nbextensions/bulkImportCell/tabs/xsvGenerator.js:37]: this gets populated when the class is instantiated using the types from the bulk import cell spec

File: nbextensions/bulkImportCell/tabs/xsvGenerator.js [hunk 3]
```diff
@@ -140,38 +127,59 @@ define(['StagingServiceClient', 'common/html', 'common/runtime', 'common/ui', 'u
             this.cssBaseClass = cssBaseClass;
         }
 
+        _setUpModel() {
```
Comment (ialarmedalien) [inline nbextensions/bulkImportCell/tabs/xsvGenerator.js:130]: separated out all this set up stuff into its own function

File: nbextensions/bulkImportCell/tabs/xsvGenerator.js [hunk 4]
```diff
@@ -140,38 +127,59 @@ define(['StagingServiceClient', 'common/html', 'common/runtime', 'common/ui', 'u
...
+                app: {
+                    spec: this.spec,
+                },
+                params: this.spec.makeDefaultedModel(),
+                inputs: {
+                    otherParamIds: this.xsvAppSpec.parameters.map((param) => {
+                        return param.id;
+                    }),
+                },
+            };
```
Comment (ialarmedalien) [inline nbextensions/bulkImportCell/tabs/xsvGenerator.js:153]: This is the model for the xsv template generator "app" -- it's used by the `paramsWidget` module to generate and validate the inputs on the XSV template form.

File: nbextensions/bulkImportCell/tabs/xsvGenerator.js [hunk 5]
```diff
@@ -140,38 +127,59 @@ define(['StagingServiceClient', 'common/html', 'common/runtime', 'common/ui', 'u
...
             const modalArgs = {
                 title: 'Create Import Specification Template',
                 body,
                 okLabel: 'Generate template!',
+                doThisFirst: doThisFirst.bind(this),
                 onConfirm: () => {
-                    // TODO: the form should monitor its state of validity and
-                    // only allow the query to be submitted when the form is valid
-                    const form = document.querySelector('#' + this.id);
-                    const params = ['output_directory', 'output_file_type', 'types'];
```
Comment (ialarmedalien) [inline nbextensions/bulkImportCell/tabs/xsvGenerator.js:154]: All this is now unnecessary as the values and state of the form are stored in `this.internalModel`, as if this were an app cell.

File: nbextensions/bulkImportCell/tabs/xsvGenerator.js [hunk 6]
```diff
@@ -140,38 +127,59 @@ define(['StagingServiceClient', 'common/html', 'common/runtime', 'common/ui', 'u
...
+            this.internalModel = Props.make({
+                data: internalModelContext,
+            });
+        }
+
         async run() {
+            this._setUpModel();
             const body = this.renderLayout();
-            const formValues = {};
+            this.startInputWidgets();
```
Comment (ialarmedalien) [inline nbextensions/bulkImportCell/tabs/xsvGenerator.js:163]: create the form

File: nbextensions/bulkImportCell/tabs/xsvGenerator.js [hunk 7]
```diff
@@ -213,6 +221,13 @@ define(['StagingServiceClient', 'common/html', 'common/runtime', 'common/ui', 'u
         renderLayout() {
             const state = this.model.getItem('state');
             let errorMessage;
+
+            const typeArray = Object.keys(state.params);
```
Comment (ialarmedalien) [inline nbextensions/bulkImportCell/tabs/xsvGenerator.js:225]: moved

File: nbextensions/bulkImportCell/tabs/xsvGenerator.js [hunk 8]
```diff
@@ -274,107 +271,103 @@ define(['StagingServiceClient', 'common/html', 'common/runtime', 'common/ui', 'u
                             ),
                         ]
                     ),
-                    // input file types
-                    fieldset([
```
Comment (ialarmedalien) [inline nbextensions/bulkImportCell/tabs/xsvGenerator.js:278]: no need for all this manual form generation with `paramsWidget` on the scene!

File: nbextensions/bulkImportCell/tabs/xsvGenerator.js [hunk 9]
```diff
@@ -274,107 +271,103 @@ define(['StagingServiceClient', 'common/html', 'common/runtime', 'common/ui', 'u
...
+                initialDisplay: this.internalModel.getItem(['inputs', 'paramDisplay']) || {},
+            });
+
+            return this.paramsWidget.start({
+                node,
+                parameters: this.spec.getSpec().parameters,
+            });
+        }
+
+        startInputWidgets() {
```
Comment (ialarmedalien) [inline nbextensions/bulkImportCell/tabs/xsvGenerator.js:312]: "Inspired by" how the configure tab sets up / sets off the params widget

File: nbextensions/bulkImportCell/tabs/xsvGenerator.js [hunk 10]
```diff
@@ -274,107 +271,103 @@ define(['StagingServiceClient', 'common/html', 'common/runtime', 'common/ui', 'u
...
+        /**
+         * Updates the stored parameter value using the bus message passed by the input widget.
+         *
+         * This is followed by evaluating the state of all inputs to see if the form is valid
+         * and ready for submission.
+         *
+         * @param {object} message - the bus message returned from the widget with a changed
+         *  parameter value
+         */
+        updateModelParameterValue(message) {
```
Comment (ialarmedalien) [inline nbextensions/bulkImportCell/tabs/xsvGenerator.js:339]: simplified version of what the configure tab uses

File: nbextensions/bulkImportCell/tabs/xsvGenerator.js [hunk 11]
```diff
@@ -274,107 +271,103 @@ define(['StagingServiceClient', 'common/html', 'common/runtime', 'common/ui', 'u
...
+            const paramIds = this.internalModel.getItem(['inputs', 'otherParamIds']),
+                paramValues = this.internalModel.getItem(['params']);
+
+            const results = await this.spec.validateParams(paramIds, paramValues, {});
+            const isValid = Object.values(results).every((param) => param.isValid);
+            this.internalModel.setItem(['state', 'params'], isValid ? 'complete' : 'incomplete');
+            const confirmButton = $(document.querySelector('#' + this.id)).nearest(
+                '[data-element="ok"]'
+            );
+            confirmButton[0].disabled = !isValid;
```
Comment (ialarmedalien) [inline nbextensions/bulkImportCell/tabs/xsvGenerator.js:368]: all important part: enable or disable the modal's confirm button, depending on whether or not the form config is valid

File: nbextensions/bulkImportCell/tabs/xsvGenerator.js [hunk 12]
```diff
@@ -398,15 +392,35 @@ define(['StagingServiceClient', 'common/html', 'common/runtime', 'common/ui', 'u
 
             types.forEach((fileType) => {
                 const appId = this.typesToFiles[fileType].appId;
+                const paramsToRemap = JSON.parse(JSON.stringify(params[fileType].params));
```
Comment (ialarmedalien) [inline nbextensions/bulkImportCell/tabs/xsvGenerator.js:395]: this ensures that the data getting sent off to the xsv template generator uses the appropriate value -- e.g. the form display value or the value in the dynamic dropdown for the species name

File: test/data/kb_uploadmethods.import_genome_as_some_crap.spec.json [hunk 13]
```diff
@@ -0,0 +1,386 @@
+{
```
Comment (ialarmedalien) [inline test/data/kb_uploadmethods.import_genome_as_some_crap.spec.json:1]: FIXME: this should not have been added!

File: test/unit/spec/nbextensions/bulkImportCell/tabs/xsvGenerator-spec.js [hunk 14]
```diff
@@ -15,6 +16,7 @@ define([
     TestUtil,
     ImportFastqInterleavedSpec,
     ImportFastqNonInterleavedSpec,
+    ImportGenbankGenomeSpec,
```
Comment (ialarmedalien) [inline test/unit/spec/nbextensions/bulkImportCell/tabs/xsvGenerator-spec.js:19]: have to add yet another spec as the existing specs don't have a sci name field

File: test/unit/spec/nbextensions/bulkImportCell/tabs/xsvGenerator-spec.js [hunk 15]
```diff
@@ -255,58 +318,12 @@ define([
...
-                    checkFormInputs(this.xsvGen, form);
-                });
-
-                it('creates a form with the appropriate inputs, multiple types', function () {
-                    this.xsvGen = createXsvGen();
-                    const form = document.createElement('div');
-                    form.innerHTML = this.xsvGen.createForm();
-                    checkFormInputs(this.xsvGen, form, true);
-                });
+            xdescribe('createForm', () => {
```
Comment (ialarmedalien) [inline test/unit/spec/nbextensions/bulkImportCell/tabs/xsvGenerator-spec.js:321]: Will add tests for form validation in the next PR

File: test/unit/spec/nbextensions/bulkImportCell/tabs/xsvGenerator-spec.js [hunk 16]
```diff
@@ -255,58 +318,12 @@ define([
...
+                // TODO!
             });
 
             describe('renderLayout', () => {
-                function checkStartUpDialog(modelData, hasError) {
+                function checkStartUpDialog(modelData, hasError, isMulti = false) {
```
Comment (ialarmedalien) [inline test/unit/spec/nbextensions/bulkImportCell/tabs/xsvGenerator-spec.js:326]: I combined the checks that had been in the old `createForm` tests into these `renderLayout` tests for simplicity

File: nbextensions/bulkImportCell/tabs/xsvGenerator.js [hunk 17]
```diff
@@ -274,107 +273,99 @@ define(['StagingServiceClient', 'common/html', 'common/runtime', 'common/ui', 'u
...
+
+            return this.paramsWidget.start({
+                node,
+                parameters: this.spec.getSpec().parameters,
+            });
+        }
+
+        startInputWidgets() {
+            this.paramsNode = document.createElement('div');
+            return Promise.try(() => {
```
Comment (briehl) [inline nbextensions/bulkImportCell/tabs/xsvGenerator.js:316]: `this.buildParamsWidget` already returns a Promise, don't need to wrap it!

File: nbextensions/bulkImportCell/tabs/xsvGenerator.js [hunk 18]
```diff
@@ -274,107 +273,99 @@ define(['StagingServiceClient', 'common/html', 'common/runtime', 'common/ui', 'u
...
+                this.buildParamsWidget(this.paramsNode).then(() => {
+                    this.updateAppConfigState();
+                });
+            });
+        }
+
+        stopInputWidgets() {
+            if (this.paramsWidget) {
+                return Promise.try(() => {
+                    this.paramsWidget.stop();
```
Comment (briehl) [inline nbextensions/bulkImportCell/tabs/xsvGenerator.js:326]: `paramsWidget.stop()` already returns a Promise.

**Review by briehl — CHANGES_REQUESTED** [review]
A couple minor things around Promise wrapping, but otherwise looks good! 👍

File: nbextensions/bulkImportCell/tabs/xsvGenerator.js [hunk 19]
```diff
@@ -274,107 +274,83 @@ define(['StagingServiceClient', 'common/html', 'common/runtime', 'common/ui', 'u
...
                 ]
             );
         }
 
+        /**
+         * This builds the parameters widget containing the inputs for the XSV form
+         */
+        startInputWidgets() {
+            this.paramsNode = document.createElement('div');
+            this.paramsBus = this.runtime
```
Comment (ialarmedalien) [inline nbextensions/bulkImportCell/tabs/xsvGenerator.js:289]: combined `startInputWidgets` with `buildInputWidgets` here

File: test/unit/spec/nbextensions/bulkImportCell/tabs/configure-spec.js [hunk 20]
```diff
@@ -51,6 +51,7 @@ define([
 
         beforeEach(() => {
             runtime = Runtime.make();
+            bus = runtime.bus();
```
Comment (ialarmedalien) [inline test/unit/spec/nbextensions/bulkImportCell/tabs/configure-spec.js:54]: moved this line up whilst looking at these tests for "inspiration" for the XSV tests

File: nbextensions/bulkImportCell/tabs/xsvGenerator.js (hunk 17 above)
Comment (ialarmedalien) [reply nbextensions/bulkImportCell/tabs/xsvGenerator.js]: Fixed

**Review by briehl — APPROVED** [review]
Looks good. Thanks for the changes!

**Pull request merged**